### ISSUE TITLE
Distributed Partd

### ISSUE THREAD

**Issue by mrocklin** [dask/dask#6164]
### Problem

I was chatting with some RAPIDS devs today (@kkraus14 @quasiben @randerzander @jakirkham @madsbk @pentschev and others ) and they were generally concerned about the scheduling overhead of dataframe merges/set_index operations (anything that uses `shuffle`).  There is an effort to improve scheduling performance generally, but there is also an option to try to rethink how we handle shuffling in dataframes, and maybe other similar operations in the future as well (like array rechunking in extreme situations).  

The currrent `"tasks"` based shuffling method uses  `O(n log(n))` tasks and log(n) full-dataset communications.  This is fine for ~100 partitions, but becomes challenging once you get to 1000 partition datasets.  We also have a separate solution around [partd](https://github.com/dask/partd), which does things in `O(n)` tasks, by buffering small shards of the data and using disk as a global communication store.  Partd as written doesn't work in a distributed fashion, but we might think about how to do some similar approach in a distributed fashion.

### DHT

In principle, building some sort of distributed hash table (DHT) like thing out of Dask parts isn't hard (actors, pubsub, etc, are all good examples of similar systems that took a couple of days to build).  What is hard is making it resilient.  For this we might look to a mixture solution between some peer-to-peer system, and a centralized `SchedulerPlugin`.  Here is a rough design.

We run tasks on each partition of the to-be-shuffled dataframe that split that partition into pieces, and put into some local splitting service.  The first task on a machine probably creates the service (we use a local thread-based-lock to coordinate between multiple tasks on the same machine).  That service splits things apart, stores them in local buffers, and periodically sends batches them to peers.  

How does it learn what peers it should send things to?  Probably it has to check in with the scheduler, which will still act as centralized control here, but won't track every individual piece of data. Maybe these things check in with the scheduler, the scheduler alerts everyone else about what is going on, and the things push data around in a way similar to a distributed hash table (DHT).  (there is lots of prior art here and algorithms to steal).  

Once all of those tasks are done we run some barrier task, and then there are a bunch of other tasks which pull output partitions from the DHT.  There is some excess communication here because we aren't thinking about locality for the Dask-DHT nodes, but for now let's just allow that as an allowable cost (my apologies).  

So great, we get by with 2*n tasks, and probably two full communications.  Not great, but it's probably better than what we have now for large shuffle computations.

### What can go wrong

So what happens if a worker goes down during execution?  We haven't replicated data, and the task that created that chunk of data is gone, so we've lost some things.  We can still recreate and rerun that task though, we just need to tell the scheduler to do so.  So in order to get resilience, we register a [`SchedulerPlugin`](https://distributed.dask.org/en/latest/plugins.html) that calls some custom code whenever a worker goes down.  We know which workers processed which tasks, so we rerun those tasks.  The other DHT nodes will get some duplicate data unfortunately, so they'll probably have to track which tasks they've seen, and know to discard any message that comes from the same task twice.  

### Nice things

So we get a nice combination here of resilience provided by the centralized scheduler, as well as low overhead provided by decentralized execution.  

If we can create this as a more general purpose DHT solution then I suspect that people will find other uses for it other than just dataframe all-to-all shuffling.

**Comment by mrocklin** [comment]
For dataframes in particular, we would make another shuffle method around this (alongside `"disk"` and `"tasks"`) that we would only use with the distributed scheduler.  It would be opt-in at first but if it performs well and robustly then it might take over as default in the future.

**Comment by kkraus14** [comment]
@mrocklin thanks for taking the time to write this up. It's much appreciated. I've made sure to circulate this to all of the relevant individuals within NVIDIA to make sure people read this and chime in their thoughts and concerns.

**Comment by mrocklin** [comment]
To be clear, this is a pretty early dump of some thoughts.  It's intended
to be the start of a conversation, not a finished proposal.

**Comment by madsbk** [comment]
@mrocklin, the `O(n log(n))` tasks and `log(n)` full-dataset communications is when using stages right? Without staging, it is `O(n**2)` tasks and `1` full-dataset communications right?

Also, can we avoid the barrier somehow? One of the lessons we learned when implementing dataframe merges using [`explicts-comms`](https://github.com/rapidsai/dask-cuda/blob/branch-0.14/dask_cuda/explicit_comms/dataframe_merge.py) was that the implicit barrier killed the overall performance and increased the memory usage significantly even thought the merge itself was much faster (@rjzamora please correct me if I am wrong). 
Which is something we also have to keep in mind when exploring approaches that uses external communication libraries.

**Comment by mrocklin** [comment]
> the O(n log(n)) tasks and log(n) full-dataset communications is when using stages right? Without staging, it is O(n**2) tasks and 1 full-dataset communications right?

Correct, up to about n==32 I think.  After that I think we start staging.  (I could have the numbers wrong here.  So when we get to partition counts like 1000 we end up with task counts in the 10000's to 100,000, but not 1,000,000

> Also, can we avoid the barrier somehow? One of the lessons we learned when implementing dataframe merges using explicts-comms was that the implicit barrier killed the overall performance and increased the memory usage significantly even thought the merge itself was much faster (@rjzamora please correct me if I am wrong).

The all-to-all nature means, I think, that computation after the merge is unlikely to start until 90% of the merge computation is done no matter what we do.  However, we can certainly do data movement during that time.  I don't think that I've seen a solution that allows for shuffling as part of a streaming workflow.

I could be wrong here though.  It sounds like you have real-world experience that says otherwise.

**Comment by madsbk** [comment]
> The all-to-all nature means, I think, that computation after the merge is unlikely to start until 90% of the merge computation is done no matter what we do.

Right, the merge itself is a loose barrier. In  [`explicts-comms`](https://github.com/rapidsai/dask-cuda/blob/branch-0.14/dask_cuda/explicit_comms/dataframe_merge.py) we have to do a barrier initially before the merge can begin, which makes it a big problem.

**Comment by mrocklin** [comment]
Right, in the approach proposed above there isn't an actuall `barrier()` call in order to start.  We have tasks that look like the following:

```python
def start_shuffle(df: pandas/cudf.DataFrame) -> None:
    shuffler = get_worker().get_shuffler()
    shuffler.put(df)
    return None

def end_shuffle(_, out_partition: int) -> pandas.cudf.DataFrame:
    shuffler = get_worker().get_shuffler()
    df = shuffler.get(out_partition)
    return df

x = ddf.map_partitions(start_shuffle)
y = some_sort_of_reduction_to_a_single_task(x)
z = expand_back_out_with(end_shuffle, ...)
```

(sorry for being very informal above.  I hope that that gets the point across)

**Comment by kkraus14** [comment]
> The currrent "tasks" based shuffling method uses O(n log(n)) tasks and log(n) full-dataset communications. This is fine for ~100 partitions, but becomes challenging once you get to 1000 partition datasets. We also have a separate solution around partd, which does things in O(n) tasks, by buffering small shards of the data and using disk as a global communication store. Partd as written doesn't work in a distributed fashion, but we might think about how to do some similar approach in a distributed fashion.

> Once all of those tasks are done we run some barrier task, and then there are a bunch of other tasks which pull output partitions from the DHT. There is some excess communication here because we aren't thinking about locality for the Dask-DHT nodes, but for now let's just allow that as an allowable cost (my apologies).

> So great, we get by with 2*n tasks, and probably two full communications. Not great, but it's probably better than what we have now for large shuffle computations.

Trading off communication complexity for task graph complexity is not a good architecture decision from my perspective. We (NVIDIA) have been pushing on things hard as we have very high performance hardware, but many people don't have this hardware. I.E. on a cloud system where networking isn't nearly as high performance as these machines we're using for benchmarking, the communication overhead is almost always significantly higher than the task graph overhead, even when we get to thousands of input partitions.

If there wasn't the task graph performance challenges we have today, we wouldn't be having this conversation as to how to shuffle data more efficiently. We'd likely be bottlenecked by our communication layer / interconnect bandwidth, which is a good bottleneck to have.

**Comment by mrocklin** [comment]
Yeah, so maybe we need to dynamically add worker location restrictions to
tasks based on the dht locations.  This would be the work of the scheduler
plugin. I would do this in a next stage of work, but maybe it's critical to
do at the outset.

**Comment by mrocklin** [comment]
cc @fjetter in case this interests you

**Comment by fjetter** [comment]
> If there wasn't the task graph performance challenges we have today, we wouldn't be having this conversation as to how to shuffle data more efficiently. We'd likely be bottlenecked by our communication layer / interconnect bandwidth, which is a good bottleneck to have.

Indeed, we (Blue Yonder) are currently bottlenecked by available memory and available network bandwidth but rarely by scheduler overhead. For this we created #6259. 
We're usually only hit by scheduler overhead if the input data is partitioned to granularly which is when we repartition (and optimize/fuse) before we start shuffling.

@crepererum made a good point there about out-of-band communication of the payload data which may tie in quite well with this approach of using a DHT (if the complexity this introduces is something we'd consider feasible)

However, pragmatically speaking, I don't think we're actually trading in communication complexity but for most scenarios keep it as is or improve it. Currently we have one full communication per stage. The number of stages we need for the reshuffling scales with `log(N)` but just to make it very clear, the number of stages is calculated using `ceil(log(N) / log(max_branch))`. To illustrate this I'll keep the constant at the default value (max_branch=32) which is probably true for most users and spell out the critical points:

| N (partitions) | stages / full comms |
|-|-|
| 10 | 1 |
| 33 | 2 |
| 1025 | 3 |

So only for relatively small dataframes with 32 partitions or less we do get away with only one data submission and this case is handled with it's dedicated code path by now (see [here](https://github.com/dask/dask/blob/622c9b453c58d81516485336dbca7a5b3a619513/dask/dataframe/shuffle.py#L471)) which is something we could still allow in the future, if need be.
For everything else, a constant two is either as bad as before or a significant improvement.

**Comment by madsbk** [comment]
I would like to point to Dask-CUDA's implementation of [shuffle](https://github.com/rapidsai/dask-cuda/blob/1d71749c6eed12ec60e175183c40300dcad914f3/dask_cuda/explicit_comms/dataframe/shuffle.py#L207), which is basically a shuffle service that perform the all-to-all communication in a single stage without the involvement of the scheduler. It works with both CUDA and regular Dataframes such as Pandas and cuDF.

In order to enable it, simply `import dask_cuda` and set `DASK_EXPLICIT_COMMS=True` or set the "explicit-comms" key in the Dask configuration and Dask will automatically use the shuffle service.

**Comment by mrocklin** [comment]
Interesting.  Have you seen substantial performance gains?

**Comment by mrocklin** [comment]
I spend 20m and put together a very rough prototype of a decentralized shuffle service using actors.

```python
from distributed.utils_test import gen_cluster

@gen_cluster(client=True)
async def test_actor_shuffle(c, s, a, b):
    import dask
    import dask.dataframe as dd
    import pandas as pd
    df = await dask.datasets.timeseries().persist()

    class Holder:
        """
        Holds shards of dataframes and then returns their accumulation

        We create one of these per output partition, and then have all of the
        input partitions send messages to them.
        """
        def __init__(self):
            self.shards = []

        def put(self, shard: pd.DataFrame) -> None:
            self.shards.append(shard)

        def get(self) -> pd.DataFrame:
            return pd.concat(self.shards, axis=0)


    holders = {char: c.submit(Holder, actor=True) for char in "abcdefghijklmnopqrstuvwxyz"}
    holders = {k: await v for k, v in holders.items()}

    def split(part: pd.DataFrame, holders=None) -> None:
        groups = part.groupby("name")
        futures = []
        for name, group in groups:
            first_letter = name.lower()[0]
            future = holders[first_letter].put(group)
            futures.append(future)
        for future in futures:
            future.result()
        return part.head(0)

    await c.compute(df.map_partitions(split, holders=holders, meta=df))

    @dask.delayed
    def get_partition(holder):
        return holder.get().result()

    out = dd.from_delayed(map(get_partition, holders.values()), meta=df).persist()

    x, y = c.compute([df, out])
    x = await x
    y = await y

    assert x.x.sum() - y.x.sum() < 1e-9
```

It works, although there are some problems

1. It's not robust.  We may want some sort of gang scheduling (this would be useful for the xgboost work as well)
2. Is passes around one actor per output partition, when we probably want one connection point per worker
3. It uses rpc, rather than something more fine-grained like batched comms (although that need may go away if we switch to one thing per worker rather than one thing per partition
4. This example isn't at all general, I've baked in that we're partitioning on the first letter of the name column


Still, I thought that this might be educational.  It shows that it doesn't have to be that hard to experiment in this space.
